<h2> Objective:- Assessment Task 2: Data Manipulation and Advanced Data Analytics for Data Science </h2>

<div style="text-align: right"> Done by: <b>Karan Murjani </b> </div>

### Common Libraries

In [1]:
import pandas as pd
import numpy as np
import json
import csv
import os
from scipy.sparse import csr_matrix
from sklearn.neighbors import NearestNeighbors
from fuzzywuzzy import process
from mlxtend.frequent_patterns import association_rules
from mlxtend.frequent_patterns import apriori
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

#### This assignment is divided into two parts. The first part focuses on data manipulation skills which include data wrangling, and the EDA and the second part focus on more advanced data analysis for data science. The first part consists of 8 questions for the provided customer order JSON data. The second part has two versions, based on the team member's student-id we are doing the second version of this part.

## Part 1 - Data Manipulation

#### Question 1.1 
#### Open the assignment2data.json file and convert it to CSV format as a data frame in pandas. The meaning of the column is in assignment2data.pdf print the shape of the CSV data-frame and find how many rows are duplicated (use pandas) and remove the duplicated rows and then print the new shape of the data frame (use pandas).

##### Answer 1.1

In [2]:
#Reading the json file
path_to_json_file = "assignment2data.json"
json_file = open(path_to_json_file, 'r')
jsonData = pd.read_json(json_file.read(), lines = True)
json_file.close()

In [3]:
#Converting json data to csv file
jsonData.to_csv('assignment2CSVData.csv', encoding = 'utf-8', index = False)

#Reading the csv file into dataframe
orderData = pd.read_csv("assignment2CSVData.csv")

In [4]:
#Finding shape of the dataframe
shape = orderData.shape
print("Shape of the dataset is = {}". format(shape))

Shape of the dataset is = (46279, 43)


In [5]:
#Finding duplicated rows 
duplicatedRowsCount = orderData.duplicated().sum()
print("Number of duplicated rows are: " + str(duplicatedRowsCount))

Number of duplicated rows are: 249


In [6]:
#Dropping duplicate values
orderData.drop_duplicates(inplace = True)

#Finding shape of the dataframe after deleting duplicates rows
newShape = orderData.shape
print("Shape of the dataset after removing duplicated rows is = {}". format(newShape))

Shape of the dataset after removing duplicated rows is = (46030, 43)


###### I have mounted the provided JSON file into the google collab working directory and then read the file. Using the pandas library I converted the JSON file into a CSV file and saved it to the same directory. The shape of the data is that we have 43 columns and 46279 rows and then checked for duplicate rows count and as a result, we got 249 rows as duplicated. After deleting the duplicate rows the new shape of the data is that we have 43 columns and 46030 rows. Yes, the solution is optimal and there is no other solution.

#### Question 1.2
#### Define a function missingdf(df) with input argument df, which could print out all the columns in data frame df and also the missing value rate for each column. For example, with a total of 1000 rows, if column1 has a missing value in 200 rows and its missing value rate will be 0:2 or 20%. The result of the function missingdf(df) will print the new data frame which has two columns: the column_name and the percent_missing.

##### Answer 2

In [7]:
#Defininig the function to find the missing value rate for each column
def missingdf(df):   
    percent_missing = df.isnull().sum() * 100 / len(df)
    missingValueDf = pd.DataFrame({'column_name': df.columns, 'percent_missing': percent_missing})
    print(missingValueDf.to_string(index = False))

#Calling the missingdf function
missingdf(orderData)

             column_name  percent_missing
              sacc_items         0.000000
             work_orders         0.000000
            female_items         0.000000
is_newsletter_subscriber         0.000000
              male_items         0.000000
       afterpay_payments         0.000000
            msite_orders         0.000000
              wftw_items         0.000000
              mapp_items         0.000000
                  orders         0.000000
             cc_payments         0.000000
             curvy_items         0.000000
         paypal_payments         0.000000
              macc_items         0.000000
                 cancels         0.000000
                 revenue         0.000000
                 returns         0.000000
 other_collection_orders         0.000000
      parcelpoint_orders         0.000000
             customer_id         0.000000
          android_orders         0.000000
   days_since_last_order         0.000000
                vouchers         0

###### We have defined the missingdf function to calculate the proportion of the missing value for each column. In the function, we have calculated the missing values based on pandas IsNull function and then divided it by the total length of the column and then added the column name and the proportion into another data frame and printed the newly created data frame. This is the best optimal solution that one can do for solving this kind of problem. 

#### Question 1.3
#### Define a function column_list(df) which could return a list which only contains the numerical column names and another list which only contains the categorical column names.
#### Use describe() function from pandas to print out the statistics for all numerical columns.

##### Answer 1.3

In [8]:
#Defininig the function to find the numerical and categorical columns list
def column_list(df):   
    allColumns = df.columns

    #Finding numerical columns name
    numericalColumns = list(df.select_dtypes([np.number]).columns)

    #Finding categorical columns name
    categoricalColumns = list(set(allColumns) - set(numericalColumns))

    return numericalColumns, categoricalColumns

numericalColumnList, categoricalColumnList = column_list(orderData)
print('-------------------------------------------------------------------')
print("Numerical column names are as below:")
print('-------------------------------------------------------------------')
print(numericalColumnList)

print('-------------------------------------------------------------------')
print('-------------------------------------------------------------------')
print("Categorical column names are as below:")
print('-------------------------------------------------------------------')
print(categoricalColumnList)

-------------------------------------------------------------------
Numerical column names are as below:
-------------------------------------------------------------------
['sacc_items', 'work_orders', 'female_items', 'male_items', 'afterpay_payments', 'msite_orders', 'wftw_items', 'mapp_items', 'orders', 'cc_payments', 'curvy_items', 'paypal_payments', 'macc_items', 'cancels', 'revenue', 'returns', 'other_collection_orders', 'parcelpoint_orders', 'android_orders', 'days_since_last_order', 'vouchers', 'average_discount_used', 'shipping_addresses', 'redpen_discount_used', 'mftw_items', 'days_since_first_order', 'unisex_items', 'home_orders', 'coupon_discount_applied', 'desktop_orders', 'ios_orders', 'apple_payments', 'wspt_items', 'wacc_items', 'items', 'mspt_items', 'devices', 'different_addresses', 'wapp_items', 'other_device_orders', 'average_discount_onoffer']
-------------------------------------------------------------------
-------------------------------------------------------

In [9]:
#Finding statistics of the numerical columns
orderData.describe(include = [np.number])

sacc_items   work_orders  female_items    male_items  \
count  46030.000000  46030.000000  46030.000000  46030.000000   
mean       0.087899      0.240995      6.472800      1.720660   
std        0.539604      1.565071     17.830007      5.468827   
min        0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      2.000000      0.000000   
75%        0.000000      0.000000      5.000000      1.000000   
max       29.000000     84.000000    537.000000    273.000000   

       afterpay_payments  msite_orders    wftw_items    mapp_items  \
count       46030.000000  46030.000000  46030.000000  46030.000000   
mean            0.053595      0.978557      1.633826      0.926613   
std             0.225220      2.859161      4.088452      3.689578   
min             0.000000      0.000000      0.000000      0.000000   
25%             0.000000      0.000000      0.000000      0.000000   
50%             0.000000      0.000000      0.000000      0.000000   
75%             0.000000      1.000000      2.000000      0.000000   
max             1.000000    172.000000    261.000000    151.000000   

             orders   cc_payments   curvy_items  paypal_payments  \
count  46030.000000  46030.000000  46030.000000     46030.000000   
mean       4.111362      0.642972      0.041321         0.490376   
std        8.122569      0.479128      0.782744         0.499913   
min        1.000000      0.000000      0.000000         0.000000   
25%        1.000000      0.000000      0.000000         0.000000   
50%        2.000000      1.000000      0.000000         0.000000   
75%        4.000000      1.000000      0.000000         1.000000   
max      665.000000      1.000000    116.000000         1.000000   

         macc_items       cancels        revenue       returns  \
count  46030.000000  46030.000000   46030.000000  46030.000000   
mean       0.573908      0.053096    1305.479232      1.626113   
std        2.542536      2.175601    5611.107402      7.019715   
min        0.000000      0.000000       0.000000      0.000000   
25%        0.000000      0.000000      76.360000      0.000000   
50%        0.000000      0.000000     213.015000      0.000000   
75%        0.000000      0.000000     791.870000      1.000000   
max      353.000000    460.000000  354700.160000    343.000000   

       other_collection_orders  parcelpoint_orders  android_orders  \
count             46030.000000        46030.000000    46030.000000   
mean                  2.308386            0.025418        0.043167   
std                   5.192695            0.452441        0.537200   
min                   0.000000            0.000000        0.000000   
25%                   1.000000            0.000000        0.000000   
50%                   1.000000            0.000000        0.000000   
75%                   2.000000            0.000000        0.000000   
max                 665.000000           32.000000       33.000000   

       days_since_last_order      vouchers  average_discount_used  \
count           46030.000000  46030.000000           46030.000000   
mean            24535.400217      0.941821            2357.118087   
std             16934.525487      2.416748            2032.366465   
min                24.000000      0.000000               0.000000   
25%              6768.000000      0.000000              80.974900   
50%             25416.000000      0.000000            2122.822500   
75%             41640.000000      1.000000            3829.608450   
max             51840.000000     57.000000           10000.000000   

       shipping_addresses  redpen_discount_used    mftw_items  \
count        46030.000000          46030.000000  46030.000000   
mean             1.501065            435.567699      0.421290   
std              1.098996           1851.597072      1.444193   
min              1.000000              0.000000      0.000000   
25%   

###### Here we defined the column_list function to perform the activities as asked. First, we collected all the columns of the data frame in a variable, then we collected numerical columns based on the data type of the columns in another variable namely ‘numericalColumns’. Finally, for finding categorical columns we just subtracted numerical columns from all columns and then printed both numerical and categorical column lists. Then we calculated statistics of only numerical columns of the data frame. There is no other way to do this problem and the solution implemented is optimal and feasible.

#### Question 1.4
#### Look into the dataframe, there are two columns which are corrupted. Could you find it out and explain the reason? (you will need to draw some visualizations or check the statistics, also you may need to look into the data to understand the meaning the columns.)

#### For the two corrupted columns, could you provide the solution to correct them?

##### Answer 1.4

In [10]:
#Corrupted column 1
print((orderData.query('shipping_addresses > 1').shape))
orderData.query('shipping_addresses > 1')[['shipping_addresses', 'different_addresses']].head()

(12521, 43)


shipping_addresses  different_addresses
0                   4                    0
1                   4                    0
2                   6                    1
7                   2                    0
8                   2                    0

In [11]:
#Corrupted column 2
print((orderData.query('days_since_last_order > days_since_first_order').shape))
orderData.query('days_since_last_order > days_since_first_order')[['days_since_last_order','days_since_first_order']].head()

(43365, 43)


days_since_last_order  days_since_first_order
0                  15672                    2091
3                  49200                    2054
4                  49272                    2053
5                  49056                    2044
6                  49008                    2042

In [12]:
#Converting into from hours to day
temp_df = orderData.copy()
temp_df['days_since_last_order_new'] = (temp_df['days_since_last_order']/24).astype(int)
print((temp_df.query('days_since_last_order_new > days_since_first_order').shape))
temp_df[['days_since_last_order_new', 'days_since_first_order']].head()

(0, 44)


days_since_last_order_new  days_since_first_order
0                        653                    2091
1                         22                    2082
2                          6                    2072
3                       2050                    2054
4                       2053                    2053

In [13]:
temp_df['days_since_last_order_new'].describe()

count    46030.000000
mean      1022.308342
std        705.605229
min          1.000000
25%        282.000000
50%       1059.000000
75%       1735.000000
max       2160.000000
Name: days_since_last_order_new, dtype: float64

In [14]:
del temp_df

In [15]:
#Corrupted column 3
print((orderData.query('revenue == 0 and orders > 0').shape))
orderData.query('revenue == 0 and orders > 0')[['customer_id', 'revenue', 'orders']].head()

(1520, 43)


customer_id  revenue  orders
4   668c6aac52ff54d4828ad379cdb38e7d      0.0       1
9   111d48b932dae281aff64cae2f17c4d6      0.0       1
18  bf193b02eae5b34ba4797ad4d3a40367      0.0       1
32  5a339c04693983ab438f112ec287bae1      0.0       1
60  18597965003161701058a5306cf74166      0.0       1

In [16]:
#Corrupted column 4
print(orderData.query('average_discount_onoffer == 0 and average_discount_used > 0').shape)
orderData.query('average_discount_onoffer == 0 and average_discount_used > 0')[['average_discount_used', 'average_discount_onoffer', 'customer_id']].head()

(3596, 43)


average_discount_used  average_discount_onoffer  \
3               3875.6715                       0.0   
22              1666.2434                       0.0   
27              1668.3017                       0.0   
32              9423.5969                       0.0   
34               625.0000                       0.0   

                         customer_id  
3   aa21f31def4edbdcead818afcdfc4d32  
22  980f142f64c8b622f8274355d4356249  
27  dbb68f4e15ec99e0657c7ebb0cd02249  
32  5a339c04693983ab438f112ec287bae1  
34  31b0695deee2d28cf3a9036c506abffe

###### There are four total columns where we can say there is data error or corrupted columns.

##### First - Based on data dictionary it is given that different_addresses is the number of times billing and shipping address was used. But if we can see from histogram, samples and statistics then we can say that different_addresses is binary column with 0 and 1 with no significance of the value. It can be human error. We can possible calculate that column by adding number of shipping address billing address. 27% of data is having this error.

##### Second - There are around 43365 samples with days_since_last_order value > days_since_first_order which shouldn't be the case. It's completely illogical. 94% of data is incorrect with this error. Here the assumption we have taken is that days_since_last_order value will not be in days but some different units for example - hours or minutes possible. We can convert it to days. After doing that we have found that there are no such samples where days_since_last_order > days_since_first_order. 

##### Third -  The ‘revenue’ column value can't be 0 there should be at least some revenue generated from the customer if he/she has placed any order. 3% of our data has this error i.e. around 1520 samples where the customer has placed an order but the revenue is 0. 

##### Fourth - The ‘average_discount_on_offer’ column value can't be 0 when the average discount used was more than 0. These two columns are correlated if the customer has used some discount then they must have got some offer discount. Close to 8% of our data has this error i.e. 3596 samples.

##### These columns are corrupted columns or data error. This can be corrected with proper domain knowledge. For example, the ‘revenue’ column value can be filled up by KNN which means that similar customers will be grouped if that customer has bought a few items it will be grouped with a similar type of customer who bought similar items.

#### Question 1.5
#### Write code and return the results by using pandas package to find out "What percentage of customers who have purchased female items have paid by credit card?”

#### Write code and return the results by using pandas package to find out "What was the total revenue to the nearest dollar for customers who have paid by credit card?”

##### Answer 1.5

In [17]:
#Finding percentage of customers who have purchased female items using credit card
femaleItemCount = len(orderData[orderData['female_items'] > 0])
ccPaymentCount = len(orderData[(orderData['female_items'] > 0) & (orderData['cc_payments'] == 1)])

percentage = round((ccPaymentCount / femaleItemCount) * 100, 2)
print('Percentage of customers who purchased female items and paid by credit card is:', percentage)

Percentage of customers who purchased female items and paid by credit card is: 65.48


In [18]:
#Finding total revenue for the customers who have purchased items using credit card
totalRevenue = round(orderData[(orderData['cc_payments'] == 1)]['revenue'].sum(), 0).astype(int)
print('The total revenue for the customers who have paid by credit card is', totalRevenue)

The total revenue for the customers who have paid by credit card is 50189329


###### The solution which we have implemented is feasible and optimal from the understanding of the language and maybe there can be another way to do it but we implemented it as per our understanding. In this, firstly we found the count of the customers who have purchased female items and then found the count of customers who has paid by credit card and purchased the female items. Finally, we found the percentage of customers who purchased female items and paid by credit card based on these two counts. In the second part of the question, we found the total revenue for the customers who have paid by credit card.

#### Question 1.6
#### Write a code to change the value of ’Y’ from column is_newsletter_subscriber to 1 and ’N’ to 0. (1 and 0 is ”int” type)

#### Print out the value count for column is_newsletter_subscriber.

##### Answer 1.6

In [19]:
#Changing the value of is_newsletter_subscriber column from Y and N to 1 and 0 resepctively
orderData['is_newsletter_subscriber_new'] = orderData['is_newsletter_subscriber'].map({'Y': 1, 'N': 0})

#Printing the dataset for viewing the new column 
orderData[['is_newsletter_subscriber', 'is_newsletter_subscriber_new']].head()

is_newsletter_subscriber  is_newsletter_subscriber_new
0                        N                             0
1                        Y                             1
2                        Y                             1
3                        Y                             1
4                        Y                             1

In [20]:
#Printing the count of value for newly created column (is_newsletter_subscriber_new)
print(orderData['is_newsletter_subscriber_new'].value_counts())

0    27203
1    18827
Name: is_newsletter_subscriber_new, dtype: int64


###### The map function can only be used to process a single value (not a list). You can split the list into different columns and use the map to replace values. Then you can integrate all the columns to achieve replacement. We used the map function here to replace yes and no with 1 and 0 respectively. And then we found the count of the values. This is the best and optimal way to replace it. We can use replace function to replace it and it works the same as the map function.

#### Question 1.7
#### Create some new features for the dataframe by using below code:
#### df [ ’ female_item_rate ’ ] = df [ ’ female_items ’ ] / df [ ’ items ’ ]
#### df [ ’ male_item_rate ’ ] = df [ ’ male_items ’ ] / df [ ’ items ’ ]
#### df [ ’ unisex_items_rate ’ ] = df [ ’ unisex_items ’ ] / df [ ’ items ’ ]

#### Write a code find out how many rows (customers) could have the value female_item_rate == 1 and the value male_item_rate == 1 and the value orders > 4.11

##### Answer 1.7

In [21]:
#Creating new features as mentioned in the task sheet - finding the item rate for gender
orderData['female_item_rate'] = orderData['female_items'] / orderData['items']
orderData['male_item_rate'] = orderData['male_items'] / orderData['items']
orderData['unisex_item_rate'] = orderData['unisex_items'] / orderData['items']

#Printing the dataset for viewing the new column 
orderData[['female_item_rate', 'male_item_rate', 'unisex_item_rate']].head()

female_item_rate  male_item_rate  unisex_item_rate
0          0.920000        0.040000          0.040000
1          0.923077        0.000000          0.076923
2          0.967105        0.019737          0.013158
3          0.000000        0.000000          1.000000
4          1.000000        0.000000          0.000000

In [22]:
#Finding the count of rows (customers) could have the value female_item_rate == 1 and the value male_item_rate == 1 and the value orders > 4.11
filteredCustomerList = orderData[(orderData['female_item_rate'] == 1) & (orderData['male_item_rate'] == 1) & (orderData['orders'] > 4.11)]['customer_id']

#printing the filtered customer data
print(filteredCustomerList)

#Finding the count of the filtered customer list
customerCount = filteredCustomerList.count()
print('The count of rows (customers) could have the value female_item_rate == 1 and the value male_item_rate == 1 and the value orders > 4.11 is:', customerCount)

Series([], Name: customer_id, dtype: object)
The count of rows (customers) could have the value female_item_rate == 1 and the value male_item_rate == 1 and the value orders > 4.11 is: 0


###### Here we created 3 new columns namely ‘Female items rate’, ‘Male items rate’ and ‘Unisex items rate’. These new columns have been created by dividing the value of total items by the values of each particular item. After that, we implemented the AND operation to find the count of customers where female_item_rate == 1 and the value male_item_rate == 1 and the value orders > 4.11 and finally print the result. We have used the AND operation as this was asked in the question and is the easiest and optimal way to solve this question. Maybe there is another way to solve this question, but as asked in the question we have used the AND operation.

#### Question 1.8
#### In this question, you will use the original format of the data to group data on the value of column is_newsletter_subscriber to show the average order value, the max order value the median order value.

##### Answer 1.8

In [23]:
#Finding the average, max and mdeain of order grouped by the value of is_newsletter_subscriber column
orderAggregation = orderData.groupby(orderData['is_newsletter_subscriber'])['orders'].agg({'mean','max','median'})

#Printing the result
orderAggregation.head()

max  median      mean
is_newsletter_subscriber                       
N                         665       1  2.484175
Y                         234       3  6.462474

###### In this, we use the older value of the is_newsletter_subscriber column to show the average order value, the max order value and the median order value. We have used the group by function to group the data based on the values of the is_newsletter_subscriber column and then we aggregated the data and found the average, max and median of the order column. There is another way to solve this question. But it takes more steps to solve this question. This is the best and optimal way to solve this question.

## Part 2 - Advanced Data Analytics for Data Science

#### Question 2.1

In [25]:
customerTransactionData = pd.read_csv('customer_transaction.csv')
print(customerTransactionData.shape)
customerTransactionData.head()

(77750, 13)


customer_id  product_id  basket_id   loyalty     household_type age_band  \
0        22115     5565932      65459  Loyalist  1 adult with kids    19-24   
1        22115     1098066      65459  Loyalist  1 adult with kids    19-24   
2        22115     1148035      65459  Loyalist  1 adult with kids    19-24   
3        22115      870882      65459  Loyalist  1 adult with kids    19-24   
4        22115     1121557      65459  Loyalist  1 adult with kids    19-24   

       department     brand               commodity  store transaction_date  \
0         Grocery  national        Breakfast snacks    374  19/01/2017 0:00   
1         Grocery   private  Baked bread/buns/rolls    374  19/01/2017 0:00   
2  Pharmaceutical  national                   Candy    374  19/01/2017 0:00   
3         Grocery  national             Frozen meat    374  19/01/2017 0:00   
4       Cosmetics   private       Coupon/misc items    374  19/01/2017 0:00   

   price  days_since_first  
0   4.99                21  
1   0.88                21  
2   0.50                21  
3   4.99                21  
4  39.99                21

In [26]:
#Checking if there is any null values.
def check_missing(df):
    null_val = df.isnull().sum()
    percent = 100 * df.isnull().sum()/len(df)
    missing_table = pd.concat([null_val, percent], axis = 1)
    col = missing_table.rename(columns = {0 : 'Num', 1 : 'Rate'})
    return col

# Display columns missing values are under 1%.
print("Number of rows: "+ str(len(customerTransactionData)))
cols = check_missing(customerTransactionData)
cols.sort_values(by = "Rate", ascending = False).head(10)

Number of rows: 77750


Num  Rate
customer_id       0   0.0
product_id        0   0.0
basket_id         0   0.0
loyalty           0   0.0
household_type    0   0.0
age_band          0   0.0
department        0   0.0
brand             0   0.0
commodity         0   0.0
store             0   0.0

In [27]:
customerTransactionData.describe()

customer_id    product_id     basket_id    store         price  \
count  7.775000e+04  7.775000e+04  7.775000e+04  77750.0  77750.000000   
mean   2.425159e+04  3.579019e+06  6.905152e+04    374.0      3.231060   
std    3.965495e+04  4.714247e+06  1.055748e+05      0.0      3.280044   
min    1.580300e+04  3.508200e+04  5.196400e+04    374.0      0.100000   
25%    1.952900e+04  9.200910e+05  5.987200e+04    374.0      1.490000   
50%    2.342900e+04  1.037863e+06  6.802100e+04    374.0      2.500000   
75%    2.699300e+04  5.578437e+06  7.521500e+04    374.0      3.990000   
max    2.811700e+06  1.820392e+07  6.811600e+06    374.0     68.970000   

       days_since_first  
count      77750.000000  
mean          33.144180  
std           43.876355  
min            0.000000  
25%            0.000000  
50%            6.000000  
75%           62.000000  
max          176.000000

In [28]:
customerTransactionData.dtypes

customer_id           int64
product_id            int64
basket_id             int64
loyalty              object
household_type       object
age_band             object
department           object
brand                object
commodity            object
store                 int64
transaction_date     object
price               float64
days_since_first      int64
dtype: object

In [29]:
customerTransactionData['loyalty'].value_counts()

Promiscuous         44943
Loyalist            31673
First Time Buyer     1134
Name: loyalty, dtype: int64

In [30]:
customerTransactionData['department'].value_counts()

Grocery           25516
Produce           16184
Pharmaceutical    11646
Meat              10867
Pastry             4648
Deli               3638
Nutrition          2472
Seafood            1494
Salad Bar           484
Cosmetics           456
Floral              345
Name: department, dtype: int64

In [31]:
#Top 5 commodity
customerTransactionData['commodity'].value_counts().head(5)

Beef              2944
Candy             2865
Tropical fruit    2148
Lunch meat        2078
Cheese            2063
Name: commodity, dtype: int64

##### How many transactions based on basket level? what is the average basket size?

In [32]:
basketLevelData = customerTransactionData.groupby(['basket_id']).agg(
    {'transaction_date' : "count"}).rename(columns = {"transaction_date" : "Transactions"}).reset_index()
print(basketLevelData.shape)
basketLevelData.head()

(3498, 2)


basket_id  Transactions
0      51964            18
1      51965            26
2      51966            25
3      51967            18
4      51968            20

In [33]:
print('Average basket size is:', int(basketLevelData['Transactions'].mean()))

Average basket size is: 22


###### Here to find the transaction on the basket level we have used the group-by pandas function and for the aggregation function, we have used the count which will generate the total transaction in each basket. There are a total of 3498 different baskets and in each basket, the average number of transactions was 22.

##### What is the most popular product commodity (based on the frequency of the purchase)?

In [34]:
#Method 1
commodityData = customerTransactionData.groupby(['commodity']).agg(
    {'commodity' : 'count'}).rename(columns = {"commodity" : "Popular_commodity"}).sort_values(by = "Popular_commodity", ascending = False).reset_index()
print(commodityData.shape)
commodityData.head()

(247, 2)


commodity  Popular_commodity
0            Beef               2944
1           Candy               2865
2  Tropical fruit               2148
3      Lunch meat               2078
4          Cheese               2063

In [35]:
#Method 2
#Top 5 commodity
customerTransactionData['commodity'].value_counts().head(5) 

Beef              2944
Candy             2865
Tropical fruit    2148
Lunch meat        2078
Cheese            2063
Name: commodity, dtype: int64

###### Here this question is solved with two methods which anyway generate the same output. Value counts in pandas generate the contribution of each category in descending order for categorical variables. So as we can see that most popular commodity based on the frequency of purchase is beef with 2944 transactions, candy with 2865 transactions, and so on.

##### What is the average of the total transaction price (average basket total price) for each customer?

In [36]:
avgBasketData = customerTransactionData.groupby(['customer_id', 'basket_id']).agg(
    {'price' : "mean"}).rename(columns =
                              {"price" : "ATV"}).sort_values(by = 'ATV', ascending = False).reset_index()
print(avgBasketData.shape)
avgBasketData.head()

(3498, 3)


customer_id  basket_id    ATV
0      2521400    6797900  15.00
1      2791100    6797700  15.00
2      2587300    6790900  15.00
3      2443300    6784900   9.99
4      2324200    6787400   9.99

###### Here group by function from pandas is used to find the average total transaction price. The average of the total price in each basket can be done by grouping based on basket_id, customer_id and the mean price is done. This is similar to ATV in business. ATV is the average transaction value and this is similar to that. This metric can be useful for businesses to evaluate ROI or determine product strategies.

##### You will need to transform the data into a format of: the row represent the basket, the column will be all product commodity, the value of the column should indicate whether the basket contains particular product commodity. Name this new dataframe as transaction_product

In [37]:
transaction_product = customerTransactionData.loc[:, ('basket_id', 'commodity')]
print(transaction_product.shape)
temp_df = pd.get_dummies(transaction_product['commodity'])
print(temp_df.shape)
transaction_product.drop(['commodity'], axis = 1, inplace = True)
transaction_product = transaction_product.join(temp_df)
del temp_df
print(transaction_product.shape)
transaction_product.head()

(77750, 2)
(77750, 247)
(77750, 248)


basket_id  Air care  Analgesics  Antacids  Apparel  Apples  \
0      65459         0           0         0        0       0   
1      65459         0           0         0        0       0   
2      65459         0           0         0        0       0   
3      65459         0           0         0        0       0   
4      65459         0           0         0        0       0   

   Audio/video products  Automotive products  Baby foods  Baby hbc  Bacon  \
0                     0                    0           0         0      0   
1                     0                    0           0         0      0   
2                     0                    0           0         0      0   
3                     0                    0           0         0      0   
4                     0                    0           0         0      0   

   Bag snacks  Baked bread/buns/rolls  Baked sweet goods  Bakery party trays  \
0           0                       0                  0                   0   
1           0                       1                  0                   0   
2           0                       0                  0                   0   
3           0                       0                  0                   0   
4           0                       0                  0                   0   

   Baking  Baking mixes  Baking needs  Bath  Bath tissues  Batteries  Beef  \
0       0             0             0     0             0          0     0   
1       0             0             0     0             0          0     0   
2       0             0             0     0             0          0     0   
3       0             0             0     0             0          0     0   
4       0             0             0     0             0          0     0   

   Beers/ales  Berries  Beverage  Bird seed  Bleach  Books  Bottle deposits  \
0           0        0         0          0       0      0                0   
1           0        0         0          0       0      0                0   
2           0        0         0          0       0      0                0   
3           0        0         0          0       0      0                0   
4           0        0         0          0       0      0                0   

   Bread  Breakfast sausage/sandwiches  Breakfast snacks  Breakfast sweets  \
0      0                             0                 1                 0   
1      0                             0                 0                 0   
2      0                             0                 0                 0   
3      0                             0                 0                 0   
4      0                             0                 0                 0   

   Broccoli/cauliflower  Brooms and mops  Butter  Cakes  Candles/accessories  \
0                     0                0       0      0                    0   
1                     0                0       0      0                    0   
2                     0                0       0      0                    0   
3                     0                0       0      0                    0   
4                     0                0       0      0                    0   

   Candy  Canned beans  Canned juices  Canned milk  Carrots  Cat food  \
0      0             0              0            0        0         0   
1      0             0              0            0        0         0   
2      1             0              0            0        0         0   
3      0             0              0            0        0         0   
4      0             0              0            0        0         0   

   Cat litter  Cereal  Cereal/breakfast  Charcoal and lighter fluid  Cheese  \
0           0       0                 0                           0       0   
1           0       0                 0                           0       0   
2           0       0                 0                           0       0   
3           0      

###### Here get_dummies function from pandas is used to do the one-hot encoding of commodity variable and then joined that with basket_id. Each row represent the basket_id and shows which commodity does that basket_id consists i.e. 1 in the column where that particular commodity is there and 0 in others. This can have multiple rows for each basket_id.
###### This kind of dataset can be used for market basket analysis by using apriori algorithm

##### You will need to transform the data into a format of: the row represent the unique customer, the column will be all product commodity, the value of the column should be the frequency of the purchase on the particular commodity cross entire data. Name this new dataframe as customer_product_freq

In [38]:
customer_product_freq = customerTransactionData.loc[:, ('customer_id', 'commodity')]
print(customer_product_freq.shape)

temp_df = pd.get_dummies(customer_product_freq['commodity'])
print(temp_df.shape)

customer_product_freq.drop(['commodity'], axis = 1, inplace = True)
customer_product_freq = customer_product_freq.join(temp_df)
del temp_df

print(customer_product_freq.shape)
customer_product_freq = customer_product_freq.groupby('customer_id').sum().reset_index()
print(customer_product_freq.shape)
customer_product_freq.head()

(77750, 2)
(77750, 247)
(77750, 248)
(3268, 248)


customer_id  Air care  Analgesics  Antacids  Apparel  Apples  \
0        15803         0           0         0        0       1   
1        15811         0           0         0        0       0   
2        15815         0           1         0        0       0   
3        15818         0           0         0        0       1   
4        15820         0           0         0        0       0   

   Audio/video products  Automotive products  Baby foods  Baby hbc  Bacon  \
0                     0                    0           0         0      0   
1                     0                    0           0         0      0   
2                     0                    0           0         0      0   
3                     0                    0           0         0      0   
4                     0                    0           0         0      2   

   Bag snacks  Baked bread/buns/rolls  Baked sweet goods  Bakery party trays  \
0           0                       1                  0                   0   
1           0                       1                  0                   0   
2           1                       2                  0                   0   
3           0                       0                  0                   0   
4           0                       0                  0                   0   

   Baking  Baking mixes  Baking needs  Bath  Bath tissues  Batteries  Beef  \
0       0             0             0     0             0          0     1   
1       0             0             1     0             0          0     1   
2       0             0             0     0             0          0     1   
3       0             0             0     0             0          0     2   
4       0             0             0     0             1          0     1   

   Beers/ales  Berries  Beverage  Bird seed  Bleach  Books  Bottle deposits  \
0           0        1         0          0       0      0                0   
1           0        0         0          0       0      0                0   
2           0        1         0          0       0      0                0   
3           0        0         0          0       0      0                0   
4           0        0         0          0       0      0                0   

   Bread  Breakfast sausage/sandwiches  Breakfast snacks  Breakfast sweets  \
0      1                             1                 0                 0   
1      0                             0                 0                 0   
2      0                             0                 0                 0   
3      0                             0                 0                 0   
4      0                             1                 0                 0   

   Broccoli/cauliflower  Brooms and mops  Butter  Cakes  Candles/accessories  \
0                     0                0       0      0                    0   
1                     0                0       0      0                    0   
2                     1                0       0      0                    0   
3                     0                0       0      0                    0   
4                     0                0       0      0                    0   

   Candy  Canned beans  Canned juices  Canned milk  Carrots  Cat food  \
0      0             0              0            0        0         0   
1      3             0              0            0        1         0   
2      1             0              0            0        0         0   
3      0             0              0            0        0         0   
4      0             0              0            0        0         0   

   Cat litter  Cereal  Cereal/breakfast  Charcoal and lighter fluid  Cheese  \
0           0       0                 0                           0       0   
1           0       0                 0                           0       0   
2           0       0                 0                           0       2   
3      

###### Here same get_dummies function is used but it is grouped with customer_id this time. Each row represents the number of times that particular commodity purchased by each customer cross entire data

###### This kind of dataset can be used for nearest neighbour aglorithm to find the similar customers to that particular customer.

##### Using the customer_product_freq to find the top 5 similar customers for each customer. (Check out the KNN)

In [39]:
#Generate sparse matrix
customer_product_freq_matrix = csr_matrix(customer_product_freq.values)
model_knn = NearestNeighbors(metric = 'cosine', algorithm = 'brute', n_neighbors = 5)
model_knn.fit(customer_product_freq_matrix)

NearestNeighbors(algorithm='brute', metric='cosine')

In [40]:
def recommender(customer_id, data,model, similar_customers):
    model.fit(data)
    idx=process.extractOne(customer_id, customer_product_freq['customer_id'])[2]
    print('Customer Selected: ', customer_product_freq['customer_id'][idx], 'Index: ', idx)
    print('----------------------------------------')
    print('Searching for similar customers.....')
    print('----------------------------------------')
    distances, indices=model.kneighbors(data[idx], n_neighbors=similar_customers + 1)
    indices = np.delete(indices, np.where(indices == idx))
    print('Top 5 similar customers for {0}:\n'.format(customer_product_freq['customer_id'][idx]))
    for i in indices:
        print(customer_product_freq['customer_id'][i])

recommender('15815', customer_product_freq_matrix, model_knn, 5) #Query the customerid

Customer Selected:  15815 Index:  2
----------------------------------------
Searching for similar customers.....
----------------------------------------
Top 5 similar customers for 15815:

26278
23233
26262
28647
19539


#### Question 2.2

##### The itemsets(basket) having length more than 1 and minimum support of 5%

In [41]:
transaction_product1 = transaction_product.groupby('basket_id').sum().reset_index()
transaction_product1['basket_id'] = transaction_product1['basket_id'].astype('str')
transaction_product1.set_index('basket_id', inplace = True)

def encode_commodity(commodity):
    if commodity <= 0:
        return 0
    if commodity >=1:
        return 1

transaction_product1 = transaction_product1.applymap(encode_commodity)
print(transaction_product1.shape)
transaction_product1.head()

(3498, 247)


Air care  Analgesics  Antacids  Apparel  Apples  \
basket_id                                                    
51964             0           0         0        0       0   
51965             0           0         0        0       1   
51966             0           0         0        0       0   
51967             0           0         0        0       0   
51968             0           0         0        0       0   

           Audio/video products  Automotive products  Baby foods  Baby hbc  \
basket_id                                                                    
51964                         0                    0           0         0   
51965                         0                    0           1         0   
51966                         0                    0           0         0   
51967                         0                    0           0         0   
51968                         0                    0           0         0   

           Bacon  Bag snacks  Baked bread/buns/rolls  Baked sweet goods  \
basket_id                                                                 
51964          0           1                       0                  0   
51965          1           0                       0                  0   
51966          0           0                       1                  0   
51967          0           0                       0                  0   
51968          0           0                       1                  0   

           Bakery party trays  Baking  Baking mixes  Baking needs  Bath  \
basket_id                                                                 
51964                       0       0             0             0     0   
51965                       0       0             0             0     0   
51966                       0       0             0             1     0   
51967                       0       0             0             0     0   
51968                       0       0             0             0     0   

           Bath tissues  Batteries  Beef  Beers/ales  Berries  Beverage  \
basket_id                                                                 
51964                 0          0     0           0        0         0   
51965                 0          0     0           0        0         0   
51966                 0          0     0           1        1         0   
51967                 0          0     0           0        0         0   
51968                 0          0     0           0        1         0   

           Bird seed  Bleach  Books  Bottle deposits  Bread  \
basket_id                                                     
51964              0       0      0                0      1   
51965              0       0      0                0      1   
51966              0       0      0                0      0   
51967              0       0      0                0      1   
51968              0       0      0                0      0   

           Breakfast sausage/sandwiches  Breakfast snacks  Breakfast sweets  \
basket_id                                                                     
51964                                 0                 0                 0   
51965                                 0                 0                 1   
51966                                 0                 0                 1   
51967                                 0                 0                 1   
51968                                 0                 0                 0   

           Broccoli/cauliflower  Brooms and mops  Butter  Cakes  \
basket_id                                                         
51964                         1                0       0      0   
51965                         0                0       0      0   
51966                         0                0       1      0   
51967                         0                0       0      0   
51968                         0                0     

In [42]:
frequent_itemsets = apriori(transaction_product1, min_support = 0.01, use_colnames=True, verbose = 1)
frequent_itemsets_filter= frequent_itemsets
print(frequent_itemsets_filter.shape)
frequent_itemsets_filter['length'] = frequent_itemsets_filter['itemsets'].apply(lambda x: len(x))
frequent_itemsets_filter = frequent_itemsets_filter.query('length >1 and support>0.05')
frequent_itemsets_filter.head()

Processing 119 combinations | Sampling itemset size 7 654
(31782, 2)


support                          itemsets  length
243  0.054317              (Apples, Bag snacks)       2
244  0.074614  (Apples, Baked bread/buns/rolls)       2
249  0.134934                    (Apples, Beef)       2
252  0.065752                   (Apples, Bread)       2
255  0.065466        (Apples, Breakfast sweets)       2

##### The association rules with minimum support of 2% and having lift more than 1.

In [43]:
#Rules
association_rules_df = association_rules(frequent_itemsets, metric = "lift", min_threshold = 1)
association_rules_df.head()

antecedents consequents  antecedent support  consequent support   support  \
0       (Candy)  (Air care)            0.550886            0.022584  0.012864   
1    (Air care)     (Candy)            0.022584            0.550886  0.012864   
2      (Cheese)  (Air care)            0.440823            0.022584  0.012293   
3    (Air care)    (Cheese)            0.022584            0.440823  0.012293   
4  (Lunch meat)  (Air care)            0.443396            0.022584  0.010863   

   confidence      lift  leverage  conviction  
0    0.023352  1.034007  0.000423    1.000786  
1    0.569620  1.034007  0.000423    1.043529  
2    0.027886  1.234744  0.002337    1.005454  
3    0.544304  1.234744  0.002337    1.227082  
4    0.024500  1.084837  0.000850    1.001964

In [44]:
print(association_rules_df[ (association_rules_df['lift'] >1) &
                    (association_rules_df['support'] >=0.02)].shape)
association_rules_df[ (association_rules_df['lift'] >1) &
                    (association_rules_df['support'] >=0.02)].head()

(30504, 9)


antecedents   consequents  antecedent support  consequent support  \
12        (Analgesics)        (Beef)            0.066038            0.557461   
13              (Beef)  (Analgesics)            0.557461            0.066038   
16             (Bread)  (Analgesics)            0.292739            0.066038   
17        (Analgesics)       (Bread)            0.066038            0.292739   
18  (Breakfast sweets)  (Analgesics)            0.295312            0.066038   

     support  confidence      lift  leverage  conviction  
12  0.038022    0.575758  1.032821  0.001208    1.043127  
13  0.038022    0.068205  1.032821  0.001208    1.002326  
16  0.020011    0.068359  1.035156  0.000680    1.002492  
17  0.020011    0.303030  1.035156  0.000680    1.014766  
18  0.021155    0.071636  1.084774  0.001653    1.006030

![Alt Text](https://media.giphy.com/media/H4zC1A2FZ0ViA6GcxD/giphy.gif)